data/src/main/java/org/apache/iceberg/data/GeneratePartitionStats.java

```diff
@@ -0,0 +1,122 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.data;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentSkipListMap;
+import org.apache.iceberg.ImmutableGenericPartitionStatisticsFile;
+import org.apache.iceberg.ManifestFile;
+import org.apache.iceberg.PartitionStatisticsFile;
+import org.apache.iceberg.PartitionStatsUtil;
+import org.apache.iceberg.Partitioning;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.types.Comparators;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GeneratePartitionStats {
+  private static final Logger LOG = LoggerFactory.getLogger(GeneratePartitionStats.class);
+
+  private final Table table;
+  private String branch;
+
+  public GeneratePartitionStats(Table table) {
+    this.table = table;
+  }
+
+  public GeneratePartitionStats(Table table, String branch) {
+    this.table = table;
+    this.branch = branch;
+  }
+
+  /**
+   * Computes the partition stats for the current snapshot and writes it into the metadata folder.
+   *
+   * @return {@link PartitionStatisticsFile} for the latest snapshot id or null if table doesn't
+   *     have any snapshot.
+   */
+  public PartitionStatisticsFile generate() {
+    Snapshot currentSnapshot = SnapshotUtil.latestSnapshot(table, branch);
+    if (currentSnapshot == null) {
+      Preconditions.checkArgument(
+          branch == null, "Couldn't find the snapshot for the branch %s", branch);
+      return null;
+    }
+
+    Types.StructType partitionType = Partitioning.partitionType(table);
+    // Map of partitionData, partition-stats-entry per partitionData.
+    // Sorting the records based on partition as per spec.
+    Map<Record, Record> partitionEntryMap =
+        new ConcurrentSkipListMap<>(Comparators.forType(partitionType));
```

Comment:
I also tried ConcurrentSkipListMap, but it seems its `compute` and `merge` methods are not guaranteed to be atomic, which can lead to consistency issues. So I guess we need to sort the records after we populate the map.

Reply:
@lirui-apache, You are right. I pushed a new test now which validates each and every stats. I observed flakiness because of  `ConcurrentSkipListMap#compute`. 

I will sort after populating the map as suggested by you and Anton. JMH, I will add in a follow up PR. I think we can incrementally improve this logic based on JMH benchmarks later on.

Reply:
Added the JMH benchmarks, for 30K manifests (1 add, 2 update operation per partition value * 10k partitions), it took only 0.5 seconds with concurrent hashmap. This includes reading 30k manifests in multi threaded way and computing the concurrent, sorting the map and writing the parquet partition stats and committing to the table. 

I wanted to try 100K partitions but the problem is generating 30k manifests and committing them to table took around 30 minutes. So, it takes lot of setup time if I increase the manifests. 

Since 0.5 seconds is very small. Do we still need to validate concurrent hashmap vs Threadlocal hashmap performance? 


```
Benchmark                                             Mode  Cnt  Score   Error  Units
PartitionStatsGeneratorBenchmark.writePartitionStats    ss    5  0.556 ± 0.343   s/op
```